plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductTemplates/AbstractProductTemplate.php

```diff
@@ -0,0 +1,145 @@
+<?php
+/**
+ * Product Block Editor abstract product template.
+ */
+
+namespace Automattic\WooCommerce\Admin\Features\ProductBlockEditor\ProductTemplates;
+
+abstract class AbstractProductTemplate {
+
+    /**
+     * Block tree root.
+     */
+    const ROOT = 'root';
+
+    /**
+     * Index for the block properties.
+     */
+    const BLOCK_PROPERTIES_INDEX = 1;
+
+    /**
+     * Index for the child blocks.
+     */
+    const BLOCK_CHILDREN_INDEX = 2;
+
+    /**
+     * Property for the internal order.
+     */
+    const ORDER_PROPERTY = '_order';
+
+    /**
+     * The block template.
+     */
+    protected $template = array();
+
+    /**
+     * Cache of reference to inserted blocks location in tree.
+     */
+    protected $cache = array();
+
+    /**
+     * Add a field to the existing template.
+     */
+    public function add_field( $field ) {
+        $args   = wp_parse_args( $field, array( 'order' => 10 ) );
+        $id     = $args['id'] ?? null;
+        $block  = $args['block'];
+        $parent = $args['parent'];
+        $order  = $args['order'];
+
+        $block[ self::BLOCK_PROPERTIES_INDEX ][ self::ORDER_PROPERTY ] = $order;
+
+        $this->insert_block( $parent, $block, $id );
+    }
+
+    /**
+     * Insert a block into the template tree.
+     */
+    private function insert_block( $parent, $block, $id ) {
+        if ( $parent === self::ROOT ) {
+            $blocks = &$this->template;
+        } else {
+            $blocks = &$this->cache[ $parent ][ self::BLOCK_CHILDREN_INDEX ];
+        }
+
+        $blocks[] = $block;
+
+        if ( $id ) {
+            $index = count( $blocks ) - 1;
+            $this->cache[ $id ] = &$blocks[ $index ];
+        }
+    }
+
+    /**
+     * Add a group to the template.
+     */
+    public function add_group( $args = array() ) {
```

Comment:
I like the idea behind this, but having "groups" and "sections" is confusing. Maybe "section" and "sub-section"? Or "group" and "sub-group"?

Reply:
Or, even just "group" or "section" and they can be nested and how they appear depends on what level they are at.

Reply:
I do like avoiding "tab", as on different devices, such as mobile, they might not look like tabs.

Reply:
Great thought!

I've toyed around with the idea of `category` and having unlimited number of nested items (or referencing them as "sections" as you mentioned above which may be better; I don't have a strong opinion on the naming yet).

On one hand, this is really great since it opens the door for unlimited nesting (which we already have multiple levels with sections in the latest template).

On the other hand, it does remove some level of intent with how the things being added should behave.  The idea of a "group" being that common items are grouped together while sections represent some kind of visual separation on the page.  But these semantics could backfire on us in the future should we change the meaning and move to something where sections/groups were identical in meaning (e.g., a flyout menu that shows groups, then sections when drilled down).

I'll post on this more in the p2 post coming Monday, but love any more thoughts/feedback you have in this area.